packages/realm-react/src/AppProvider.tsx

```diff
@@ -62,20 +58,84 @@ type AppProviderProps = Realm.AppConfiguration & {
   children: React.ReactNode;
 };
 
+type AppProviderWithAppProps = {
+  app: Realm.App;
+  children: React.ReactNode;
+};
+
+type AppProviderProps = AppProviderWithAppProps & AppProviderWithConfigurationProps;
+type DynamicAppProviderWithAppProps = RestrictivePick<AppProviderProps, keyof AppProviderWithAppProps>;
+type DynamicAppProviderWithConfigurationProps = RestrictivePick<
+  AppProviderProps,
+  keyof AppProviderWithConfigurationProps
+>;
+
+/**
+ * Props for the AppProvider component. You can either pass an existing app through the `appInstance` prop
+ * or props that replicate the configuration that is used to create a Realm.App instance:
+ * https://www.mongodb.com/docs/realm-sdks/js/latest/Realm.App.html#~AppConfiguration
+ */
+type DynamicAppProviderProps = DynamicAppProviderWithAppProps | DynamicAppProviderWithConfigurationProps;
+
+/**
+ * React component providing a Realm App instance on the context for the
+ * sync hooks to use. An `AppProvider` is required for an app to use the hooks.
+ * @param props - Either the {@link Realm.AppConfiguration} for App Services passed as props **or** the {@link Realm.App} passed through the `app` prop.
+ * @param appRef - Provides a ref to the app instance, which can be used to access the app instance outside of the React component tree. **Not available when using the `app` prop**.
+ */
+export function AppProvider(props: DynamicAppProviderProps): React.ReactNode;
 /**
  * React component providing a Realm App instance on the context for the
  * sync hooks to use. An `AppProvider` is required for an app to use the hooks.
- * @param appProps - The {@link Realm.AppConfiguration} for app services, passed as props.
+ * @param appProps - The {@link Realm.AppConfiguration} for App Services, passed as props.
  * @param appRef - A ref to the app instance, which can be used to access the app instance outside of the React component tree.
  */
-export const AppProvider: React.FC<AppProviderProps> = ({ children, appRef, ...appProps }) => {
-  const configuration = useRef<Realm.AppConfiguration>(appProps);
+export function AppProvider(props: DynamicAppProviderWithConfigurationProps): React.ReactNode;
+/**
+ * React component providing a Realm App instance on the context for the
+ * sync hooks to use. An `AppProvider` is required for an app to use the hooks.
+ * @param appInstance - The {@link Realm.App} for the provider.
+ */
+export function AppProvider(props: DynamicAppProviderWithAppProps): React.ReactNode;
+export function AppProvider({ children, app, ...appWithConfigurationProps }: DynamicAppProviderProps): React.ReactNode {
+  if (app instanceof App) {
+    if (Object.keys(appWithConfigurationProps).length > 0) {
+      throw new Error("Cannot use configuration props when using an existing App instance.");
+    }
+    return <AppProviderWithApp app={app}>{children}</AppProviderWithApp>;
+  } else if (app != null) {
```

Comment:
I'd like for us to avoid loose (not) equality.

```suggestion
  } else if (typeof app === "object") {
```

Reply:
in this context, isn't more loose what we want? This kind of assumes that maybe they use JavaScript so ended up passing anything, so could be even not an object, I'd generally not have this but since we have this breaking change and until now they could have gotten away with anything in the app field, seemed necessary. Otherwise object is actually disallowed by TypeScript. I could do `!app`.